pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +212,174 @@ pub mod pallet {
+				.bonded_currencies
+				.iter()
+				.map(|id| (id.clone(), T::Fungibles::total_issuance(id.clone())))
+				.filter(|(_, iss)| iss.gt(&Zero::zero()))
+				.collect();
+
+			let mut sum_of_issuances: FungiblesBalanceOf<T> = total_issuances
+				.iter()
+				.fold(Zero::zero(), |sum, (_, x)| sum.saturating_add(*x));
+
+			ensure!(!sum_of_issuances.is_zero(), Error::<T>::ZeroAmount);
+
+			let mut dead_currencies = Vec::with_capacity(total_issuances.len());
+			for (currency_id, token_issuance) in total_issuances.iter() {
+				let burnt = T::Fungibles::decrease_balance(
+					currency_id.clone(),
+					&who,
+					Bounded::max_value(),
+					Precision::BestEffort,
+					Preservation::Expendable,
+					Fortitude::Force,
+				)?;
+
+				let amount = burnt
+					.checked_mul(&total_collateral_issuance)
+					.ok_or(ArithmeticError::Overflow)? // TODO: do we need a fallback if this fails?
+					.checked_div(&sum_of_issuances)
+					.unwrap_or(Zero::zero());
+
+				T::CollateralCurrency::transfer(
+					T::CollateralAssetId::get(),
+					&pool_account,
+					&who,
+					amount,
+					Preservation::Expendable,
+				)?;
+
+				total_collateral_issuance -= amount;
+				sum_of_issuances -= burnt;
+
+				// if the total issuance drops to 0 due to this burn, kill the currency
+				if token_issuance <= &burnt {
+					dead_currencies.push(currency_id);
+				}
+			}
+
+			// destroy all active currencies if the collateral locked in the pool has dropped to 0
+			if total_collateral_issuance.is_zero() || sum_of_issuances.is_zero() {
+				// we assume all currencies that already had their total issuance at 0 have already been deactivated in a previous step
+				for (currency_id, _) in total_issuances {
+					T::Fungibles::start_destroy(currency_id, None)?;
+				}
+			} else {
+				// deactivate all currencies whose issuance dropped to 0 in this step
+				for currency_id in dead_currencies {
+					T::Fungibles::start_destroy(currency_id.clone(), None)?;
+				}
+			}
+
+			Ok(())
 		}
 
 		#[pallet::call_index(9)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn finish_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn finish_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::InUse);
+
+			for currency_id in pool_details.bonded_currencies {
+				if T::Fungibles::asset_exists(currency_id.clone()) {
+					T::Fungibles::finish_destroy(currency_id.clone())?;
+				}
+			}
+
+			let pool_account = pool_id.clone().into();
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			T::CollateralCurrency::transfer(
+				T::CollateralAssetId::get(),
+				&pool_account,
+				&pool_details.manager,
+				total_collateral_issuance,
+				Preservation::Expendable,
+			)?;
```

Comment:
If the pool account is already destroyed by the `refund_account` tx, then this command would fail right? So the pool can never be deleted

Reply:
Could be, I can add a check if it exists. Will investigate if a 0 amount transfer fails if the account is non-existent.

Reply:
Or else I just make sure that the refunding does not kill the account?

Reply:
Another solution would be, that we take the deposit from the collateral, depending on if we want to use the deposit pallet.

Reply:
hmm wait, we touched the pool account right? So this means the account should not get dusted.

Reply:
15369478fd8d5b498cdb7b59f0941bf5bee7fca8

Reply:
O amount transfers should always succeed from what I've seen, but I added conditionals anyway to avoid performing needless lookups.